### PR TITLE
Dev items

### PR DESCRIPTION
**Pull request by RlonRyan**
Fixed item rendering, and corrected bad auto-imports.

### CODE REVIEW

File: src/main/java/com/infinityraider/infinitylib/item/IInfinityItem.java [hunk 1]
```diff
@@ -3,9 +3,16 @@
...
 public interface IInfinityItem extends IToggleable {
-    public String getInternalName();
 
-    List<String> getOreTags();
+	String getInternalName();
+
+	List<String> getOreTags();
+	
+	@SideOnly(Side.CLIENT)
+	void registerItemRenderer();
```
Comment (InfinityRaider) [inline src/main/java/com/infinityraider/infinitylib/item/IInfinityItem.java]: Item rendering shouldn't be handled inside the item class, it should be done similarly to block rendering.

File: src/main/java/com/infinityraider/infinitylib/render/block/IBlockRenderingHandler.java [hunk 2]
```diff
@@ -67,6 +67,10 @@
     void renderWorldBlock(ITessellator tessellator, World world, BlockPos pos, double x, double y, double z,
                           IBlockState state, B block, @Nullable T tile, boolean dynamicRender, float partialTick, int destroyStage);
 
+	default void renderDynamic(ITessellator tess, T tile){};
+	
+	default void renderStatic(ITessellator tess, IBlockState state){};
+
```
Comment (InfinityRaider) [inline src/main/java/com/infinityraider/infinitylib/render/block/IBlockRenderingHandler.java]: Why?

Comment (RlonRyan) [inline src/main/java/com/infinityraider/infinitylib/render/block/IBlockRenderingHandler.java]: It's cleaner and removes unneeded if statements testing if the mode is dynamic or static... The intention is to completely replace the renderWorldBlock method.

File: src/main/java/com/infinityraider/infinitylib/item/IInfinityItem.java (hunk 1 above)
Comment (RlonRyan) [inline src/main/java/com/infinityraider/infinitylib/item/IInfinityItem.java]: Item rendering isn't... It's just a renderer registration hook...

**Comment by RlonRyan** [comment]
Is this good for a merge?

File: src/main/java/com/infinityraider/infinitylib/render/block/IBlockRenderingHandler.java (hunk 2 above)
Comment (InfinityRaider) [inline src/main/java/com/infinityraider/infinitylib/render/block/IBlockRenderingHandler.java]: Fine, but then pass all other arguments too, static rendering can use tile info too, for example the plant on crops, the wood type, etc.
World should also be passed, pos as well, if interaction with neighbour blocks is necessary.
IBlockState has to be passed in both as well.
partialtick only for dynamic renders because it is useful in interpolation for animations.

File: src/main/java/com/infinityraider/infinitylib/item/IInfinityItem.java (hunk 1 above)
Comment (InfinityRaider) [inline src/main/java/com/infinityraider/infinitylib/item/IInfinityItem.java]: But then this method is called to register it, i assume?
I don't think this method is necessary because the  ModHelper class can register items from all the information provided from other methods.

**Comment by InfinityRaider** [comment]
Not yet.

File: src/main/java/com/infinityraider/infinitylib/render/block/IBlockRenderingHandler.java (hunk 2 above)
Comment (RlonRyan) [inline src/main/java/com/infinityraider/infinitylib/render/block/IBlockRenderingHandler.java]: The static really should only get the tessellator and the blockstate, as it shouldn't be looking up dynamic properties. The pos and world are contained in the blockstate and the tileentity, so I thought I'd cleanup the method by dropping them from the parameters. Also, if you have the tile entity then why bother with the blockstate?

I guess the best approach would be to mimic the normal TESR draw args for the dynamic render...

Comment (InfinityRaider) [inline src/main/java/com/infinityraider/infinitylib/render/block/IBlockRenderingHandler.java]: dynamic = redraw vertices every tick
static = redraw vertices every chunk update.

Why should a static render (i.e. something which renders the same for long periods of time) not have access to block state and tile entity data?

Comment (InfinityRaider) [inline src/main/java/com/infinityraider/infinitylib/render/block/IBlockRenderingHandler.java]: Block state might contain data not stored in the tile entity, such as plant growth stage.

Comment (RlonRyan) [inline src/main/java/com/infinityraider/infinitylib/render/block/IBlockRenderingHandler.java]: But the tile entity has access to both the block and the meta.

Comment (RlonRyan) [inline src/main/java/com/infinityraider/infinitylib/render/block/IBlockRenderingHandler.java]: Static rendering has access to the blockstate.
It doesn't have access to the tile entity, since it would wind up being a dummy tile entity, or a tile entity that could be invalid as soon as the next tick, meaning that the static render would have to be called again to be displayed correctly.

Comment (InfinityRaider) [inline src/main/java/com/infinityraider/infinitylib/render/block/IBlockRenderingHandler.java]: Why would it be a dummy/invalid tile? The code works.

Comment (RlonRyan) [inline src/main/java/com/infinityraider/infinitylib/render/block/IBlockRenderingHandler.java]: It works because behind the scenes you pass the itemstack to the dummy tile and reload the dummy tile.

Comment (InfinityRaider) [inline src/main/java/com/infinityraider/infinitylib/render/block/IBlockRenderingHandler.java]: What do you mean ItemStack? This is block rendering, not inventory

File: src/main/java/com/infinityraider/infinitylib/render/block/BlockRenderer.java [hunk 3]
```diff
@@ -175,106 +170,8 @@ public ItemCameraTransforms getItemCameraTransforms() {
...
-        private final B block;
-        private final T tile;
-        private final ItemStack stack;
-        private final World world;
-        private final EntityLivingBase entity;
-        private ItemCameraTransforms.TransformType transformType;
-        private final VertexFormat format;
-        private final Function<ResourceLocation, TextureAtlasSprite> bakedTextureGetter;
-
-        private BakedItemModel(World world, B block, T tile, ItemStack stack, EntityLivingBase entity, IBlockRenderingHandler<B, T> renderer, VertexFormat format, Function<ResourceLocation, TextureAtlasSprite> bakedTextureGetter) {
```
Comment (InfinityRaider) [inline src/main/java/com/infinityraider/infinitylib/render/block/BlockRenderer.java]: why is all of thsi gone?

File: src/main/java/com/infinityraider/infinitylib/render/block/IBlockRenderingHandler.java (hunk 2 above)
Comment (InfinityRaider) [inline src/main/java/com/infinityraider/infinitylib/render/block/IBlockRenderingHandler.java]: I'm not sure where you get this 'dummy tile' from, but what i do is the following:

I store a pointer to the tile entity ehre: https://github.com/InfinityRaider/InfinityLib/blob/master/src/main/java/com/infinityraider/infinitylib/block/BlockTileCustomRenderedBase.java#L20

Then retrieve it here: https://github.com/InfinityRaider/InfinityLib/blob/master/src/main/java/com/infinityraider/infinitylib/render/block/BlockRenderer.java#L118

File: src/main/java/com/infinityraider/infinitylib/render/block/BlockRenderer.java (hunk 3 above)
Comment (RlonRyan) [inline src/main/java/com/infinityraider/infinitylib/render/block/BlockRenderer.java]: It was duplicate code, that was covered by the item renderer.

Comment (InfinityRaider) [inline src/main/java/com/infinityraider/infinitylib/render/block/BlockRenderer.java]: ah i see, ye that makes sense.

File: src/main/java/com/infinityraider/infinitylib/block/ICustomRenderedBlock.java [hunk 4]
```diff
@@ -11,17 +9,18 @@
 import net.minecraftforge.fml.relauncher.SideOnly;
 
 import java.util.List;
+import com.infinityraider.infinitylib.render.block.RenderBlock;
 
 /**
  * Implemented in a Block class to have special rendering handling for the block
  */
-public interface ICustomRenderedBlock<T extends TileEntityBase> {
+public interface ICustomRenderedBlock {
```
Comment (InfinityRaider) [inline src/main/java/com/infinityraider/infinitylib/block/ICustomRenderedBlock.java]: Why are the generics gone? They were useful that you did not have to cast the TileEntity to its actual type anymore.

File: src/main/java/com/infinityraider/infinitylib/item/ICustomRenderedItem.java [hunk 5]
```diff
@@ -1,34 +1,15 @@
...
     @SideOnly(Side.CLIENT)
-    IItemRenderingHandler<I> getRenderer();
-
-    /**
-     * Gets a list of all textures necessary to render this item,
-     * returned textures will be stitched to the atlas
-     * @return list of textures
-     */
-    @SideOnly(Side.CLIENT)
-    List<ResourceLocation> getTextures();
```
Comment (InfinityRaider) [inline src/main/java/com/infinityraider/infinitylib/item/ICustomRenderedItem.java]: why is this gone? Textures need to be stitched.

File: src/main/java/com/infinityraider/infinitylib/item/IItemWithModel.java [hunk 6]
```diff
@@ -1,19 +0,0 @@
...
-import net.minecraft.util.Tuple;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
-
-import java.util.List;
-
-/**
- * Interface used to ease model registering
- */
-public interface IItemWithModel extends IInfinityItem {
```
Comment (InfinityRaider) [inline src/main/java/com/infinityraider/infinitylib/item/IItemWithModel.java]: why did you remove this?

File: src/main/java/com/infinityraider/infinitylib/render/RenderUtilBase.java [hunk 7]
```diff
@@ -31,21 +31,21 @@
 public abstract class RenderUtilBase {
     protected RenderUtilBase() {}
 
-    public int getMixedBrightness(IBlockAccess world, BlockPos pos, Block block) {
+    public static int getMixedBrightness(IBlockAccess world, BlockPos pos, Block block) {
```
Comment (InfinityRaider) [inline src/main/java/com/infinityraider/infinitylib/render/RenderUtilBase.java]: I specifically made these none static because all my renderers extend this class.

File: src/main/java/com/infinityraider/infinitylib/render/block/BlockRenderer.java [hunk 8]
```diff
@@ -1,277 +1,46 @@
...
-import org.lwjgl.opengl.GL11;
 
-import javax.vecmath.Matrix4f;
 import java.util.*;
 
 @SideOnly(Side.CLIENT)
-public class BlockRenderer<B extends BlockBase & ICustomRenderedBlock<T>, T extends TileEntityBase> extends TileEntitySpecialRenderer<T> implements IModel {
-    private final B block;
-    private final IBlockRenderingHandler<B, T> renderer;
+public class BlockRenderer<B extends BlockBase & ICustomRenderedBlock> implements IModel {
```
Comment (InfinityRaider) [inline src/main/java/com/infinityraider/infinitylib/render/block/BlockRenderer.java]: The TileEntity generic was really useful, why did you remove it?

File: src/main/java/com/infinityraider/infinitylib/render/block/BlockRenderer.java [hunk 9]
```diff
@@ -1,277 +1,46 @@
...
+		this.renderer = renderer;
+	}
 
-    @Override
-    public Collection<ResourceLocation> getTextures() {
-        return renderer.getAllTextures();
-    }
+	@Override
+	public Collection<ResourceLocation> getDependencies() {
+		return Collections.emptyList();
```
Comment (InfinityRaider) [inline src/main/java/com/infinityraider/infinitylib/render/block/BlockRenderer.java]: Again, textures have to be stitched....

File: src/main/java/com/infinityraider/infinitylib/render/tessellation/TessellatorAbstractBase.java [hunk 10]
```diff
@@ -25,690 +25,774 @@
...
+				break;
+			case COLOR:
+				data[0] = getRedValueFloat();
+				data[1] = getGreenValueFloat();
+				data[2] = getBlueValueFloat();
+				data[3] = getAlphaValueFloat();
+				break;
+		}
+		return data;
+	}
```
Comment (InfinityRaider) [inline src/main/java/com/infinityraider/infinitylib/render/tessellation/TessellatorAbstractBase.java]: Why does it mark this entire block as changed when nothing has changed?

Comment (RlonRyan) [inline src/main/java/com/infinityraider/infinitylib/render/tessellation/TessellatorAbstractBase.java]: What? I don't remember touching that...

File: src/main/java/com/infinityraider/infinitylib/render/RenderUtilBase.java (hunk 7 above)
Comment (RlonRyan) [inline src/main/java/com/infinityraider/infinitylib/render/RenderUtilBase.java]: But what about renderers that can't extend the class?

Comment (InfinityRaider) [inline src/main/java/com/infinityraider/infinitylib/render/RenderUtilBase.java]: all rendering base classes are interfaces or extend this alraedy.
In the other case a static instance field will be fine.

File: src/main/java/com/infinityraider/infinitylib/item/IInfinityItem.java (hunk 1 above)
Comment (RlonRyan) [inline src/main/java/com/infinityraider/infinitylib/item/IInfinityItem.java]: This method is called by the ModHelper class to allow for more configurability from an item's perspective.

File: src/main/java/com/infinityraider/infinitylib/item/ICustomRenderedItem.java (hunk 5 above)
Comment (RlonRyan) [inline src/main/java/com/infinityraider/infinitylib/item/ICustomRenderedItem.java]: That's handled by the renderer itself.

File: src/main/java/com/infinityraider/infinitylib/item/IItemWithModel.java (hunk 6 above)
Comment (RlonRyan) [inline src/main/java/com/infinityraider/infinitylib/item/IItemWithModel.java]: I moved it to the registerRenderers() hook.

File: src/main/java/com/infinityraider/infinitylib/render/RenderUtilBase.java (hunk 7 above)
Comment (RlonRyan) [inline src/main/java/com/infinityraider/infinitylib/render/RenderUtilBase.java]: The subclasses still inherit the static methods, its just now other classes can use them too.

File: src/main/java/com/infinityraider/infinitylib/render/block/BlockRenderer.java (hunk 8 above)
Comment (RlonRyan) [inline src/main/java/com/infinityraider/infinitylib/render/block/BlockRenderer.java]: It's still there, but just in a sub-interface for blocks with tiles.
This interface is for blocks that may or may not have tiles associated with them to render.

File: src/main/java/com/infinityraider/infinitylib/item/IInfinityItem.java (hunk 1 above)
Comment (InfinityRaider) [inline src/main/java/com/infinityraider/infinitylib/item/IInfinityItem.java]: Oh okay, I missed that. 
Nevermind

File: src/main/java/com/infinityraider/infinitylib/item/ICustomRenderedItem.java (hunk 5 above)
Comment (InfinityRaider) [inline src/main/java/com/infinityraider/infinitylib/item/ICustomRenderedItem.java]: Makes sense to move the textures to the render class instead of the block class indeed.

**Comment by RlonRyan** [comment]
So...
